### PR TITLE
NLP visualization

### PR DESCRIPTION
**Pull request by KathrynElrod**

### CODE REVIEW

File: src/components/Fhir/InsightsDetailButton.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                <Popup
+                    trigger={
+                        <button>
+                            <i className="fa fa-lightbulb-o fas fa-bold"/>
+                        </button>
+                    }
+                    position="left center"
+                    on={['hover', 'click', "focus"]}
+                >
+                    <div style={{boxShadow: "0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)"}} className="panel panel-default">
```
Comment (cragun47) [inline src/components/Fhir/InsightsDetailButton.js:46]: Inline style.

File: src/components/Fhir/InsightsDetailButton.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                        <div className="panel-heading">
+                            <b className="text-primary">
+                                <i className="fa fa-lightbulb-o fas fa-bold" style={{paddingRight: 10}}/>Insights for {rec.resourceType} {rec.id}
+                            </b>
+                        </div>
+                        <div className="table-responsive">
+                            <table className="table table-condensed table-hover table-striped table-bordered">
```
Comment (cragun47) [inline src/components/Fhir/InsightsDetailButton.js:53]: You probably don't have accessibility as a requirement, but an easy add is If the table is being used for layout only, add  role="presentation".  (https://fae.disability.illinois.edu/rulesets/TABLE_5/)

File: src/components/Fhir/ResourceList.js [hunk 3]
```diff
@@ -40,7 +51,7 @@ function renderCell(record, allOf, oneOf) {
                 <tbody>
                     {
                         entries.map((o, i) => (
-                            <tr key={i}>
+                            <tr key={i} className={o.isNLP ? "mark" : ""}>
```
Comment (cragun47) [inline src/components/Fhir/ResourceList.js:54]: Better style changing using ternary then adding / removing a class.   Good job.

File: src/components/Fhir/ResourceList.js [hunk 4]
```diff
@@ -377,6 +405,30 @@ export default class ResourceList extends React.Component
             ])
         });
 
+        // 4. Insights ---------------------------------------------------------
+        out.push ({
+            label: <div style={{textAlign: 'center'}}>
+                <button
+                    title="Highlight NLP Codes"
+                    onMouseUp={ this.toggleHighlight }
+                    style={{
```
Comment (cragun47) [inline src/components/Fhir/ResourceList.js:414]: Change inline styles to .css classes.

File: src/lib/index.js [hunk 5]
```diff
@@ -390,6 +390,116 @@ export function getCodeOrConcept(data, defaultValue = "-") {
...
+    var result = {}
+
+    // get data.meta.extension.*.extension.*.url
+    let meta = getPath(data, "meta");
+    if (meta) {
+        result.lastUpdated = getPath(meta, "lastUpdated")
+        let ext_outer = getPath(meta, "extension")
+        if (ext_outer && Array.isArray(ext_outer)) {
+            for (let item_outer in ext_outer) {
+                if (getPath(ext_outer[item_outer], "url") == "http://ibm.com/fhir/cdm/insight/result") {
```
Comment (cragun47) [inline src/lib/index.js:423]: Very deep conditional and looping logic.  Should refactor to digestible subroutines.  Use the subroutine names to document what is going on.  Example:

```javascript
 for (let item_inner in ext_inner) {
                             let url = getPath(ext_inner[item_inner], "url")
                             // now that we've got the url, find all the different bits
                             result = gatherInsightsForInnerItem(result, url)
                             }
```

**Review by cragun47 — CHANGES_REQUESTED** [review]
Remove confidential content; move inline styles to classes; reduce complex logic blocks.   See comments.

Good work!

File: src/components/Fhir/InsightsDetailButton.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+export default class InsightsDetailButton extends React.Component {
+    static PropTypes = {
+        settings: PropTypes.object.isRequired,
+        resource: PropTypes.object
+    }
+
+    render() {
+        let rec = this.props.resource
+        let deets = getInsightDetails(rec)
```
[the remark this answers is not in the document]
Comment (KathrynElrod) [reply src/components/Fhir/InsightsDetailButton.js:17]: Renamed to `details`.

File: src/components/Fhir/InsightsDetailButton.js (hunk 2 above)
Comment (KathrynElrod) [reply src/components/Fhir/InsightsDetailButton.js:53]: accessibility should always be a requirement -- thanks for the tip!

File: build/js/report.html [hunk 7]
```diff
@@ -3,12 +3,40 @@
...
+/**
+ * Carrot Search FoamTree HTML5 (demo variant)
+ * v3.4.5, 4fa198d722d767b68d0409e88290ea6de98d1eaa/4fa198d7, build FOAMTREE-SOFTWARE4-DIST-39, Jul 26, 2017
+ * 
+ * Carrot Search confidential.
+ * Copyright 2002-2017, Carrot Search s.c, All Rights Reserved.
+ */
```
[the remark this answers is not in the document]
Comment (KathrynElrod) [reply build/js/report.html:34]: I believe this is a report file generated for the developer, and is not included in the built application. 
image

File: src/components/Fhir/Grid/index.js [hunk 8]
```diff
@@ -48,14 +48,12 @@ export class Grid extends React.Component
...
+            url += "?_format=json&_pretty=true"
         }
 
         return (
-            <tr
-                key={i}
-                onClick={ () => window.open(url, "_blank") }
-                style={{ cursor: "pointer" }}
-            >
+            <tr key={i}>
```
[the remark this answers is not in the document]
Comment (KathrynElrod) [reply src/components/Fhir/Grid/index.js:56]: This is not code added by me -- rather, I removed the `onClick` and `style` parameters and condensed the block into a single line.

File: build/js/report.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (cragun47) [reply build/js/report.html:34]: Why is it in the HTML?   IMO, if we are going to keep it, it is something that I think Eric should be made aware of and approve.

File: src/components/Fhir/Grid/index.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (cragun47) [reply src/components/Fhir/Grid/index.js:56]: Good solution!

File: src/components/Fhir/InsightsDetailButton.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                            <i className="fa fa-lightbulb-o fas fa-bold"/>
+                        </button>
+                    }
+                    position="left center"
+                    on={['hover', 'click', "focus"]}
+                >
+                    <div className="drop-shadow panel panel-default">
+                        <div className="panel-heading">
+                            <b className="text-primary">
+                                <i className="fa fa-lightbulb-o fas fa-bold" style={{paddingRight: 10}}/>Insights for {rec.resourceType} {rec.id}
```
Comment (cragun47) [inline src/components/Fhir/InsightsDetailButton.js:48]: Move inline styles to class.

File: src/components/Fhir/ResourceList.js [hunk 10]
```diff
@@ -40,8 +53,8 @@ function renderCell(record, allOf, oneOf) {
                 <tbody>
                     {
                         entries.map((o, i) => (
-                            <tr key={i}>
-                                <td className="label-cell">{ o.label }</td>
+                            <tr key={i} className={o.isHighlight ? "bg-primary" : ""}>
+                                <td className="label-cell" style={o.isHighlight ? {color: "white"} : null}>{ o.label }</td>
```
Comment (cragun47) [inline src/components/Fhir/ResourceList.js:57]: Recommend move inline style to class.

**Review by cragun47 — APPROVED** [review]
Approved.  Minor recommendations.

**Review by ewill00 — APPROVED** [review]
LGTM - you/Brian @cragun47 can fill me in on one of his comments but otherwise I'm good with delivering it.

**Pull request merged**